src/jobs/ega/axios/idpClient.ts

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) 2024 The Ontario Institute for Cancer Research. All rights reserved
+ *
+ * This program and the accompanying materials are made available under the terms of
+ * the GNU Affero General Public License v3.0. You should have received a copy of the
+ * GNU Affero General Public License along with this program.
+ *  If not, see <http://www.gnu.org/licenses/>.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import axios, { AxiosError } from 'axios';
+import jwt from 'jsonwebtoken';
+import urlJoin from 'url-join';
+import { getAppConfig } from '../../../config';
+import logger from '../../../logger';
+import getAppSecrets from '../../../secrets';
+
+import { EGA_GRANT_TYPE, EGA_REALMS_PATH, EGA_TOKEN_ENDPOINT } from '../../../utils/constants';
+import { IdpToken } from '../types/responses';
+const { verify } = jwt;
+
+const CLIENT_NAME = 'IDP_CLIENT';
+
+/**
+ * Verifies an access token string has valid signature + is not expired
+ * Uses jsonwebtoken.verify
+ * @param token
+ * @returns jwt.JwtPayload | undefined | string
+ */
+const decodeToken = async (
+  token: string,
+): Promise<jwt.JwtPayload | 'TokenExpiredError' | undefined> => {
+  logger.info('Verifying token');
+  const {
+    auth: { egaPublicKey },
+  } = await getAppSecrets();
+
+  const decoded = verify(token, egaPublicKey, { algorithms: ['RS256'] });
+  if (typeof decoded == 'string' || decoded === null) {
+    switch (true) {
+      case decoded === 'TokenExpiredError':
+        return 'TokenExpiredError';
+      case decoded === 'JsonWebTokenError':
+        logger.error(`Invalid JWT format`);
+        return undefined;
+      default:
+        logger.error(`Error decoding JWT`);
+        return undefined;
+    }
+  }
+  return decoded;
+};
+
+/**
+ * Uses jsonwebtoken.verify to validate token is not expired
+ * @param token IdpToken
+ * @returns Promise<boolean>
+ */
+export const tokenExpired = async (token: IdpToken): Promise<boolean> => {
```

Comment:
Convention is probably to name this `isTokenExpired` but this is a minor quibble. The description should probably state somethings like "returns true if token is expired, otherwise returns false. the token may be invalid and still return false"